src/controller/java/AndroidLogDownloadFromNode.cpp

```diff
@@ -0,0 +1,292 @@
+
+    auto * self = static_cast<AndroidLogDownloadFromNode *>(context);
+    VerifyOrReturn(self != nullptr, ChipLogProgress(Controller, "Send command failure callback with null context. Ignoring"));
+
+    self->FinishLogDownloadFromNode(err);
+}
+
+void AndroidLogDownloadFromNode::FinishLogDownloadFromNode(CHIP_ERROR err)
+{
+    CHIP_ERROR jniErr = CHIP_NO_ERROR;
+    if (mBdxReceiver != nullptr)
+    {
+        if (mTimeout > 0)
+        {
+            mBdxReceiver->CancelBDXTransferTimeout();
+        }
+        delete mBdxReceiver;
+        mBdxReceiver = nullptr;
+    }
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+    JniLocalReferenceScope scope(env);
+
+    if (err == CHIP_NO_ERROR)
+    {
+        ChipLogProgress(Controller, "Log Download succeeded.");
+        jmethodID onSuccessMethod;
+        // Java method signature : boolean onSuccess(int fabricIndex, long nodeId)
+        jniErr = JniReferences::GetInstance().FindMethod(env, mJavaCallback.ObjectRef(), "onSuccess", "(IJ)V", &onSuccessMethod);
+
+        VerifyOrReturn(jniErr == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find onSuccess method"));
+
+        env->CallVoidMethod(mJavaCallback.ObjectRef(), onSuccessMethod, static_cast<jint>(mController->GetFabricIndex()),
+                            static_cast<jlong>(mRemoteNodeId));
+        return;
+    }
+
+    ChipLogError(Controller, "Log Download Failed : %" CHIP_ERROR_FORMAT, err.Format());
+
+    jmethodID onErrorMethod;
+    // Java method signature : void onError(int fabricIndex, long nodeId, long errorCode)
+    jniErr = JniReferences::GetInstance().FindMethod(env, mJavaCallback.ObjectRef(), "onError", "(IJJ)V", &onErrorMethod);
+    VerifyOrReturn(jniErr == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find onError method"));
+
+    env->CallVoidMethod(mJavaCallback.ObjectRef(), onErrorMethod, static_cast<jint>(mController->GetFabricIndex()),
+                        static_cast<jlong>(mRemoteNodeId), static_cast<jlong>(err.AsInteger()));
+}
+
+void AndroidLogDownloadFromNode::OnBdxTransferCallback(void * context, FabricIndex fabricIndex, NodeId remoteNodeId,
+                                                       const chip::ByteSpan & data, CHIP_ERROR * errInfoOnFailure)
+{
+    auto * self = static_cast<AndroidLogDownloadFromNode *>(context);
+    VerifyOrReturn(self != nullptr, ChipLogProgress(Controller, "Send command failure callback with null context. Ignoring"));
+
+    self->OnTransferCallback(fabricIndex, remoteNodeId, data, errInfoOnFailure);
+}
+
+void AndroidLogDownloadFromNode::OnTransferCallback(FabricIndex fabricIndex, NodeId remoteNodeId, const chip::ByteSpan & data,
+                                                    CHIP_ERROR * errInfoOnFailure)
+{
+    VerifyOrReturn(mJavaCallback.HasValidObjectRef(), ChipLogError(Controller, "mJavaCallback is invalid"));
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
```

Comment:
please check if env is nullptr or not, thanks

Reply:
Modified. Thanks.